adr/002-Custom-Resource-Deserialization-Problem.md

```diff
@@ -0,0 +1,37 @@
+# Multi Version Custom Resources Deserialization Problem
+
+## Status
+
+accepted
+
+## Context
+
+In case there are multiple versions of a custom resource in can happen that a controller/informer tracking
+such a resource might run into deserialization problem as shown
+in [this integration test](https://github.com/java-operator-sdk/java-operator-sdk/blob/07aab1a9914d865364d7236e496ef9ba5b50699e/operator-framework/src/test/java/io/javaoperatorsdk/operator/MultiVersionCRDIT.java#L55-L55)
+. In the mentioned case two versions of a custom resource are not compatible with each other. The informer receives
+both, but naturally not able to deserialize one of them.
```

Comment:
Can we clarify somewhere in the doc that the informer/watcher receives an event for resource with a given Group Version Kind (GVK) but with the structure of a different (former or newer) version (for the same Group and Kind).

This is what in the end might lead to a deserialization Exception if the `Serializer` is configured in a **strict** mode and some of the fields might have unmatched target types because there was a change in between the two versions of the same resource definition.

Note that the deserializer can be set to be more **lenient** by configuring the Serialization Unmatched Field Type module:
```java
Serialization.UNMATCHED_FIELD_TYPE_MODULE.setRestrictToTemplates(true);
```

Reply:
> Note that the deserializer can be set to be more lenient by configuring the Serialization Unmatched Field Type module:

I think this is something we want to avoid. So processing resource that are not complete in some cases, that can lead to some more confusion. But might be good to put this into the docs as a note.

Reply:
> Can we clarify somewhere in the doc that the informer/watcher receives an event for resource with a given Group Version Kind (GVK) but with the structure of a different (former or newer) version (for the same Group and Kind).

added some explanation, not sure if you meant it this way